<a href="https://colab.research.google.com/github/akshaygaur2407/Akshay/blob/main/modeling.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Model Building**

In [ ]:
#importing required packages from sklearn 
!pip3 install catboost
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.svm import SVC
from sklearn.linear_model import LogisticRegressionCV
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import ExtraTreesClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_val_score
from sklearn.ensemble import AdaBoostClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.model_selection import GridSearchCV
from sklearn.ensemble import VotingClassifier
from sklearn.naive_bayes import GaussianNB
from xgboost import XGBClassifier
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score, precision_score 
from catboost import CatBoostClassifier

In [ ]:
#importing datasets
all_data=pd.read_csv('/content/all_data.csv')  # all_data 
labels=pd.read_csv('/content/0bf8bc6e-30d0-4c50-956a-603fc693d966.csv') # labels for train data
train=pd.read_csv('/content/4910797b-ee55-40a7-8668-10efd5c1b960.csv')  # training data
test=pd.read_csv('/content/702ddfc5-68cd-4d1d-a0de-f5f566f76d91.csv') # test data

In [ ]:
all_data = all_data.drop(['id', 'subvillage', 'lga', 'ward','recorded_by'], axis=1) #dropping irrelevent columns
categorical = ['funder',
               'installer',
               'basin',
               'scheme_management',
               'extraction_type_class',
               'management',
               'management_group',
               'payment_type',
               'water_quality',
               'quantity',
               'source',
               'source_class',
               'waterpoint_type','permit','quantity_group']
for i in categorical:
    all_data[i] = pd.factorize(all_data[i])[0] #fatorzing categorical data in zeros and one

In [ ]:
labels = pd.factorize(labels.status_group)[0] #factorzing label column 

In [ ]:
labels[:20]

array([0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 2, 0])

In [ ]:
len(test)

14850

In [ ]:
test.head()

id  amount_tsh  ...     waterpoint_type waterpoint_type_group
0  50785         0.0  ...               other                 other
1  51630         0.0  ...  communal standpipe    communal standpipe
2  17168         0.0  ...               other                 other
3  45559         0.0  ...               other                 other
4  49871       500.0  ...  communal standpipe    communal standpipe

[5 rows x 40 columns]

In [ ]:
all_data.head()

Unnamed: 0  funder  gps_height  ...  population_log   age  season
0           0       0      1390.0  ...        4.700480  12.0     2.0
1           1       1      1399.0  ...        5.638355   3.0     2.0
2           2       2       686.0  ...        5.525453   4.0     1.0
3           3       3       263.0  ...        4.077537  27.0     1.0
4           4       4      1300.0  ...        5.525453  14.0     3.0

[5 rows x 28 columns]

In [ ]:
all_data=all_data.drop(['Unnamed: 0','origin'],axis=1)

In [ ]:
train_X= all_data[0:len(train)]
test_X = all_data[len(train):]

In [ ]:
train_X.head()

funder  gps_height  installer  ...  population_log   age  season
0       0      1390.0          0  ...        4.700480  12.0     2.0
1       1      1399.0          1  ...        5.638355   3.0     2.0
2       2       686.0          2  ...        5.525453   4.0     1.0
3       3       263.0          3  ...        4.077537  27.0     1.0
4       4      1300.0          4  ...        5.525453  14.0     3.0

[5 rows x 26 columns]

In [ ]:
# Dividing data into training and testing data
X_train,X_test,y_train,y_test=train_test_split(train_X,labels,test_size=.3,random_state=69) 

In [ ]:
# training logisticregressionCV model 
model_lr=LogisticRegressionCV(cv=5)
model_lr.fit(X_train,y_train)
prediction_lr=model_lr.predict(X_test)
accuracy_score(prediction_lr,y_test)

/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:818: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG,
/usr/local/lib/python3.7/dist-packages/sklearn/linear_model/_logistic.py:818: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression


0.6450617283950617

In [ ]:
# training baseline CatBoostClassifier model on training data
model_cbc = CatBoostClassifier()
model_cbc.fit(X_train,y_train)
prediction_cbc=model_cbc.predict(X_test)
accuracy_score(y_test,prediction_cbc)
#0.7948372615039282

Learning rate set to 0.095694
0:	learn: 1.0339412	total: 23.1ms	remaining: 23s
1:	learn: 0.9826706	total: 43ms	remaining: 21.5s
2:	learn: 0.9401278	total: 62.2ms	remaining: 20.7s
3:	learn: 0.9055528	total: 82.2ms	remaining: 20.5s
4:	learn: 0.8773003	total: 101ms	remaining: 20.1s
5:	learn: 0.8526779	total: 121ms	remaining: 20.1s
6:	learn: 0.8316219	total: 141ms	remaining: 20.1s
7:	learn: 0.8134272	total: 161ms	remaining: 19.9s
8:	learn: 0.7975711	total: 181ms	remaining: 19.9s
9:	learn: 0.7839744	total: 204ms	remaining: 20.2s
10:	learn: 0.7719648	total: 223ms	remaining: 20s
11:	learn: 0.7621246	total: 242ms	remaining: 19.9s
12:	learn: 0.7535751	total: 263ms	remaining: 19.9s
13:	learn: 0.7455481	total: 281ms	remaining: 19.8s
14:	learn: 0.7382398	total: 301ms	remaining: 19.8s
15:	learn: 0.7315624	total: 320ms	remaining: 19.7s
16:	learn: 0.7256337	total: 340ms	remaining: 19.7s
17:	learn: 0.7193852	total: 364ms	remaining: 19.9s
18:	learn: 0.7144534	total: 384ms	remaining: 19.8s
19:	learn: 0.

0.7948372615039282

In [ ]:
# training baseline RandomForestClassifier model on training data
model_rf=RandomForestClassifier()
model_rf.fit(X_train,y_train)
prediction_rf=model_rf.predict(X_test)
accuracy_score(prediction_rf,y_test)

0.7997755331088664

In [ ]:
# using cross_val_score to cross validate randomforest model 
score = cross_val_score(model_rf, train_X, labels, scoring='accuracy', cv=3).mean()
print('3 fold cross-validation score:', score.round(4))

3 fold cross-validation score: 0.7973


In [ ]:
#training basline xgboost model 
xgb = XGBClassifier(max_depth=10, n_estimators=300, silent=False, booster='gbtree')

# Fit the model
xgb.fit(X_train, y_train)

# make predictions for test data
y_pred = xgb.predict(X_test)

# evaluate predictions
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred, average='weighted')

# Print the results
print("Accuracy: %.2f%%" % (accuracy * 100.0))
print("Precision: %.2f%%" % (precision * 100.0))

Accuracy: 80.88%
Precision: 80.36%


In [ ]:
score = cross_val_score(xgb, train_X, labels, scoring='accuracy', cv=3).mean()
print('3 fold cross-validation score:', score.round(4))

3 fold cross-validation score: 0.8026


In [ ]:
#training baseline extratree model
model_etc=ExtraTreesClassifier()
model_etc.fit(X_train,y_train)
y_pred=model_etc.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred, average='weighted')

# Print the results
print("Accuracy: %.2f%%" % (accuracy * 100.0))
print("Precision: %.2f%%" % (precision * 100.0))

Accuracy: 78.86%
Precision: 78.28%


Here three algorithms namely xgboost, randomforest,catboost give us the highest accuracy

Now, we will do some hyperparameter tuning to improve our score further 

In [ ]:
xgb1 = XGBClassifier(max_depth=10, n_estimators=300, silent=False, booster='gbtree')

# Fit the model
xgb1.fit(X_train, y_train)

# make predictions for test data
y_pred = xgb1.predict(X_test)

# evaluate predictions
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred, average='weighted')

# Print the results
print("Accuracy: %.2f%%" % (accuracy * 100.0))
print("Precision: %.2f%%" % (precision * 100.0))

Accuracy: 80.88%
Precision: 80.36%


In [ ]:
print(xgb1)

XGBClassifier(max_depth=10, n_estimators=300, objective='multi:softprob',
              silent=False)


In [ ]:
# tuning n_estimator
for i in range(250,400,50): 
  xgb1 = XGBClassifier(max_depth=10, n_estimators=i, silent=False, booster='gbtree')

# Fit the model
  xgb1.fit(X_train, y_train)

# make predictions for test data
  y_pred = xgb1.predict(X_test)

# evaluate predictions
  accuracy = accuracy_score(y_test, y_pred)
  precision = precision_score(y_test, y_pred, average='weighted')

# Print the results
  print('For n_estimator ' i)
  print("Accuracy: %.2f%%" % (accuracy * 100.0))
  print("Precision: %.2f%%" % (precision * 100.0))

For n_estimator  250.00%
Accuracy: 80.81%
Precision: 80.31%
For n_estimator  300.00%
Accuracy: 80.88%
Precision: 80.36%
For n_estimator  350.00%
Accuracy: 80.80%
Precision: 80.25%


In [ ]:
#tuning max depth
for i in range(5,12): 
  xgb1 = XGBClassifier(max_depth=i, n_estimators=300, silent=False, booster='gbtree')

# Fit the model
  xgb1.fit(X_train, y_train)

# make predictions for test data
  y_pred = xgb1.predict(X_test)

# evaluate predictions
  accuracy = accuracy_score(y_test, y_pred)
  precision = precision_score(y_test, y_pred, average='weighted')

# Print the results
  print('For n_estimator  ' i)
  print("Accuracy: %.2f%%" % (accuracy * 100.0))
  print("Precision: %.2f%%" % (precision * 100.0))

For n_estimator  5.00%
Accuracy: 78.87%
Precision: 78.78%
For n_estimator  6.00%
Accuracy: 79.98%
Precision: 79.81%
For n_estimator  7.00%
Accuracy: 80.29%
Precision: 79.91%
For n_estimator  8.00%
Accuracy: 80.59%
Precision: 80.13%
For n_estimator  9.00%
Accuracy: 80.82%
Precision: 80.35%
For n_estimator  10.00%
Accuracy: 80.88%
Precision: 80.36%
For n_estimator  11.00%
Accuracy: 80.49%
Precision: 79.92%


In [ ]:
# tuning min_child_weight
for i in range(4,13,2): 
  xgb1 = XGBClassifier(max_depth=10, n_estimators=300, silent=False, booster='gbtree',min_child_weight=i)

# Fit the model
  xgb1.fit(X_train, y_train)

# make predictions for test data
  y_pred = xgb1.predict(X_test)

# evaluate predictions
  accuracy = accuracy_score(y_test, y_pred)
  precision = precision_score(y_test, y_pred, average='weighted')

# Print the results
  print('For min_child_weight 'i)
  print("Accuracy: %.2f%%" % (accuracy * 100.0))
  print("Precision: %.2f%%" % (precision * 100.0))

For min_child_weight 4.00%
Accuracy: 80.72%
Precision: 80.21%
For min_child_weight 6.00%
Accuracy: 80.72%
Precision: 80.21%
For min_child_weight 8.00%
Accuracy: 80.87%
Precision: 80.39%
For min_child_weight 10.00%
Accuracy: 80.89%
Precision: 80.41%
For min_child_weight 12.00%
Accuracy: 80.70%
Precision: 80.20%


In [ ]:
# tuning gamma values
for i in range(1,6,1): 
xgb1 = XGBClassifier(max_depth=10, n_estimators=300, silent=False, booster='gbtree',min_child_weight=10,gamma=i/10)

# Fit the model
  xgb1.fit(X_train, y_train)

# make predictions for test data
  y_pred = xgb1.predict(X_test)

# evaluate predictions
  accuracy = accuracy_score(y_test, y_pred)
  precision = precision_score(y_test, y_pred, average='weighted')

# Print the results
  print('For min_child_weight ' i)
  print("Accuracy: %.2f%%" % (accuracy * 100.0))
  print("Precision: %.2f%%" % (precision * 100.0))

For min_child_weight 1.00%
Accuracy: 80.74%
Precision: 80.26%
For min_child_weight 2.00%
Accuracy: 80.79%
Precision: 80.33%
For min_child_weight 3.00%
Accuracy: 80.79%
Precision: 80.31%
For min_child_weight 4.00%
Accuracy: 80.92%
Precision: 80.47%
For min_child_weight 5.00%
Accuracy: 80.89%
Precision: 80.41%


In [ ]:
# tuning reg_aplha value
params=[.1]
for i in params: 
  xgb1 = XGBClassifier(max_depth=10, n_estimators=300, silent=True, booster='gbtree',min_child_weight=10,gamma=.5,reg_alpha=i)

# Fit the model
  xgb1.fit(X_train, y_train)

# make predictions for test data
  y_pred = xgb1.predict(X_test)

# evaluate predictions
  accuracy = accuracy_score(y_test, y_pred)
  precision = precision_score(y_test, y_pred, average='weighted')

# Print the results
  print('For reg_alpha ' i)
  print("Accuracy: %.2f%%" % (accuracy * 100.0))
  print("Precision: %.2f%%" % (precision * 100.0))

For reg_alpha 0.10%
Accuracy: 80.95%
Precision: 80.52%


The best accuracy with XGBoost is 80.95% 
max_depth=10, n_estimators=300, silent=False, booster='gbtree',min_child_weight=10,gamma=.5,reg_alpha=i

Tunning randomforest model

In [ ]:
#tuning n_estimator value 
params=[100,200,300,400,500,600,700]
for i in (params):
  model=RandomForestClassifier(n_estimators=i)
  model.fit(X_train,y_train)
  y_pred = model.predict(X_test)

# evaluate predictions
  accuracy = accuracy_score(y_test, y_pred)
  precision = precision_score(y_test, y_pred, average='weighted')

# Print the results
  print('For n_estimator 'i)
  print("Accuracy: %.2f%%" % (accuracy * 100.0))
  print("Precision: %.2f%%" % (precision * 100.0))

For reg_alpha 100.00%
Accuracy: 79.99%
Precision: 79.36%
For reg_alpha 200.00%
Accuracy: 79.90%
Precision: 79.26%
For reg_alpha 300.00%
Accuracy: 80.07%
Precision: 79.42%
For reg_alpha 400.00%
Accuracy: 80.07%
Precision: 79.45%
For reg_alpha 500.00%
Accuracy: 79.99%
Precision: 79.34%
For reg_alpha 600.00%
Accuracy: 80.12%
Precision: 79.49%
For reg_alpha 700.00%
Accuracy: 80.06%
Precision: 79.41%


In [ ]:
# tuning max_depth value
params=[15,16,18,20,22]
for i in (params):
  model=RandomForestClassifier(n_estimators=500,max_depth=i)
  model.fit(X_train,y_train)
  y_pred = model.predict(X_test)

# evaluate predictions
  accuracy = accuracy_score(y_test, y_pred)
  precision = precision_score(y_test, y_pred, average='weighted')

# Print the results
  print('For max depth ' i)
  print("Accuracy: %.2f%%" % (accuracy * 100.0))
  print("Precision: %.2f%%" % (precision * 100.0))

For reg_alpha 15.00%
Accuracy: 79.85%
Precision: 80.15%
For reg_alpha 16.00%
Accuracy: 80.01%
Precision: 79.99%
For reg_alpha 18.00%
Accuracy: 80.42%
Precision: 80.14%
For reg_alpha 20.00%
Accuracy: 80.49%
Precision: 80.04%
For reg_alpha 22.00%
Accuracy: 80.35%
Precision: 79.77%


In [ ]:
 #tuned catboostclassifier values
 clf = CatBoostClassifier(iterations=1000, custom_metric=['AUC','Accuracy'], loss_function='MultiClass',
                         bootstrap_type='Bernoulli', boosting_type='Plain', 
                         max_depth=8, random_state=42, od_type="Iter")
 clf.fit(X_train,y_train)
 y_pred=clf.predict(X_test)
 accuracy_score(y_pred,y_test)

Learning rate set to 0.095694
0:	learn: 1.0327277	total: 61.2ms	remaining: 1m 1s
1:	learn: 0.9795809	total: 126ms	remaining: 1m 2s
2:	learn: 0.9368813	total: 186ms	remaining: 1m 1s
3:	learn: 0.9014708	total: 250ms	remaining: 1m 2s
4:	learn: 0.8702963	total: 313ms	remaining: 1m 2s
5:	learn: 0.8451478	total: 376ms	remaining: 1m 2s
6:	learn: 0.8223901	total: 436ms	remaining: 1m 1s
7:	learn: 0.8039581	total: 500ms	remaining: 1m 1s
8:	learn: 0.7873829	total: 561ms	remaining: 1m 1s
9:	learn: 0.7731050	total: 627ms	remaining: 1m 2s
10:	learn: 0.7606087	total: 697ms	remaining: 1m 2s
11:	learn: 0.7491163	total: 762ms	remaining: 1m 2s
12:	learn: 0.7390370	total: 822ms	remaining: 1m 2s
13:	learn: 0.7297741	total: 890ms	remaining: 1m 2s
14:	learn: 0.7198236	total: 953ms	remaining: 1m 2s
15:	learn: 0.7125958	total: 1.02s	remaining: 1m 2s
16:	learn: 0.7047812	total: 1.09s	remaining: 1m 2s
17:	learn: 0.6988834	total: 1.15s	remaining: 1m 2s
18:	learn: 0.6934698	total: 1.21s	remaining: 1m 2s
19:	learn:

0.8011223344556678

In [ ]:
from sklearn.svm import SVC
from sklearn.ensemble import StackingClassifier

In [ ]:
# stacking three best performing models
from sklearn.ensemble import StackingClassifier
from sklearn.model_selection import RepeatedStratifiedKFold
from sklearn.model_selection import cross_val_score
def get_stacking():
	# define the base models
	level0 = list()
	level0.append(('rf',RandomForestClassifier(n_estimators=500,max_depth=20)))
	level0.append(('cat',CatBoostClassifier(iterations=1000, loss_function='MultiClass',
                         bootstrap_type='Bernoulli', boosting_type='Plain', 
                         max_depth=8, random_state=42, od_type="Iter")))
	level0.append(('xgb', XGBClassifier(max_depth=10, n_estimators=300, silent=True, booster='gbtree',min_child_weight=10,gamma=.5,reg_alpha=.1)))
	# define meta learner model
	level1 = SVC()
	# define the stacking ensemble
	model = StackingClassifier(estimators=level0, final_estimator=level1, cv=3)
	return model

# get a list of models to evaluate
def get_models():
	models = dict()
	models['stacking'] = get_stacking()
	return models

# evaluate a give model using cross-validation
def evaluate_model(model, X, y):
	scores = cross_val_score(model, X, y, scoring='accuracy', cv=3, n_jobs=-1, error_score='raise')
	return scores

# get the models to evaluate
models = get_models()
# evaluate the models and store results
results, names = list(), list()
for name, model in models.items():
	scores = evaluate_model(model, X_train, y_train)
	results.append(scores)
	names.append(name)
	print(name, scores.mean())

stacking 0.801058201058201


In [ ]:
model

StackingClassifier(cv=3,
                   estimators=[('rf',
                                RandomForestClassifier(max_depth=20,
                                                       n_estimators=500)),
                               ('cat',
                                <catboost.core.CatBoostClassifier object at 0x7ff537c1bcd0>),
                               ('xgb',
                                XGBClassifier(gamma=0.5, max_depth=10,
                                              min_child_weight=10,
                                              n_estimators=300, reg_alpha=0.1,
                                              silent=True))],
                   final_estimator=SVC())

In [ ]:
model.fit(X_train,y_train)

Learning rate set to 0.095694
0:	learn: 1.0327277	total: 69.4ms	remaining: 1m 9s
1:	learn: 0.9795809	total: 132ms	remaining: 1m 5s
2:	learn: 0.9368813	total: 200ms	remaining: 1m 6s
3:	learn: 0.9014708	total: 260ms	remaining: 1m 4s
4:	learn: 0.8702963	total: 319ms	remaining: 1m 3s
5:	learn: 0.8451478	total: 381ms	remaining: 1m 3s
6:	learn: 0.8223901	total: 445ms	remaining: 1m 3s
7:	learn: 0.8039581	total: 505ms	remaining: 1m 2s
8:	learn: 0.7873829	total: 570ms	remaining: 1m 2s
9:	learn: 0.7731050	total: 629ms	remaining: 1m 2s
10:	learn: 0.7606087	total: 692ms	remaining: 1m 2s
11:	learn: 0.7491163	total: 751ms	remaining: 1m 1s
12:	learn: 0.7390370	total: 812ms	remaining: 1m 1s
13:	learn: 0.7297741	total: 872ms	remaining: 1m 1s
14:	learn: 0.7198236	total: 935ms	remaining: 1m 1s
15:	learn: 0.7125958	total: 1.01s	remaining: 1m 1s
16:	learn: 0.7047812	total: 1.07s	remaining: 1m 1s
17:	learn: 0.6988834	total: 1.13s	remaining: 1m 1s
18:	learn: 0.6934698	total: 1.19s	remaining: 1m 1s
19:	learn:

StackingClassifier(cv=3,
                   estimators=[('rf',
                                RandomForestClassifier(max_depth=20,
                                                       n_estimators=500)),
                               ('cat',
                                <catboost.core.CatBoostClassifier object at 0x7ff537c1bcd0>),
                               ('xgb',
                                XGBClassifier(gamma=0.5, max_depth=10,
                                              min_child_weight=10,
                                              n_estimators=300, reg_alpha=0.1,
                                              silent=True))],
                   final_estimator=SVC())

In [ ]:
y_pred=model.predict(X_test)

In [ ]:
accuracy_score(y_test,y_pred)

0.8078563411896745

In [ ]:
model.fit(X_test,y_test)

Learning rate set to 0.091647
0:	learn: 1.0386240	total: 41.8ms	remaining: 41.8s
1:	learn: 0.9879705	total: 74.7ms	remaining: 37.3s
2:	learn: 0.9480356	total: 108ms	remaining: 35.9s
3:	learn: 0.9136051	total: 141ms	remaining: 35.2s
4:	learn: 0.8869357	total: 175ms	remaining: 34.8s
5:	learn: 0.8606375	total: 213ms	remaining: 35.2s
6:	learn: 0.8390780	total: 246ms	remaining: 34.9s
7:	learn: 0.8200596	total: 280ms	remaining: 34.7s
8:	learn: 0.8030695	total: 314ms	remaining: 34.6s
9:	learn: 0.7871843	total: 349ms	remaining: 34.6s
10:	learn: 0.7737284	total: 383ms	remaining: 34.4s
11:	learn: 0.7614232	total: 419ms	remaining: 34.5s
12:	learn: 0.7516837	total: 452ms	remaining: 34.3s
13:	learn: 0.7419573	total: 485ms	remaining: 34.2s
14:	learn: 0.7339212	total: 519ms	remaining: 34.1s
15:	learn: 0.7243619	total: 552ms	remaining: 34s
16:	learn: 0.7170784	total: 586ms	remaining: 33.9s
17:	learn: 0.7099760	total: 623ms	remaining: 34s
18:	learn: 0.7027119	total: 656ms	remaining: 33.9s
19:	learn: 0.

StackingClassifier(cv=3,
                   estimators=[('rf',
                                RandomForestClassifier(max_depth=20,
                                                       n_estimators=500)),
                               ('cat',
                                <catboost.core.CatBoostClassifier object at 0x7ff537c1bcd0>),
                               ('xgb',
                                XGBClassifier(gamma=0.5, max_depth=10,
                                              min_child_weight=10,
                                              n_estimators=300, reg_alpha=0.1,
                                              silent=True))],
                   final_estimator=SVC())

In [ ]:
model2=StackingClassifier(cv=3,
                   estimators=[('rf',
                                RandomForestClassifier(max_depth=20,
                                                       n_estimators=500)),
                               ('cat',
                                CatBoostClassifier(iterations=1000, loss_function='MultiClass',
                         bootstrap_type='Bernoulli', boosting_type='Plain', 
                         max_depth=8, random_state=42, od_type="Iter")),
                               ('xgb',
                                XGBClassifier(gamma=0.5, max_depth=10,
                                              min_child_weight=10,
                                              n_estimators=300, reg_alpha=0.1,
                                              silent=True))],
                   final_estimator=XGBClassifier())

In [ ]:
model2.fit(X_train,y_train)

Learning rate set to 0.095694
0:	learn: 1.0327277	total: 44.6ms	remaining: 44.6s
1:	learn: 0.9795809	total: 90.8ms	remaining: 45.3s
2:	learn: 0.9368813	total: 132ms	remaining: 43.8s
3:	learn: 0.9014708	total: 173ms	remaining: 43s
4:	learn: 0.8702963	total: 217ms	remaining: 43.2s
5:	learn: 0.8451478	total: 258ms	remaining: 42.8s
6:	learn: 0.8223901	total: 297ms	remaining: 42.2s
7:	learn: 0.8039581	total: 339ms	remaining: 42s
8:	learn: 0.7873829	total: 379ms	remaining: 41.7s
9:	learn: 0.7731050	total: 422ms	remaining: 41.8s
10:	learn: 0.7606087	total: 466ms	remaining: 41.9s
11:	learn: 0.7491163	total: 506ms	remaining: 41.6s
12:	learn: 0.7390370	total: 546ms	remaining: 41.4s
13:	learn: 0.7297741	total: 586ms	remaining: 41.2s
14:	learn: 0.7198236	total: 631ms	remaining: 41.4s
15:	learn: 0.7125958	total: 671ms	remaining: 41.2s
16:	learn: 0.7047812	total: 710ms	remaining: 41s
17:	learn: 0.6988834	total: 753ms	remaining: 41.1s
18:	learn: 0.6934698	total: 792ms	remaining: 40.9s
19:	learn: 0.68

StackingClassifier(cv=3,
                   estimators=[('rf',
                                RandomForestClassifier(max_depth=20,
                                                       n_estimators=500)),
                               ('cat',
                                <catboost.core.CatBoostClassifier object at 0x7fb859777dd0>),
                               ('xgb',
                                XGBClassifier(gamma=0.5, max_depth=10,
                                              min_child_weight=10,
                                              n_estimators=300, reg_alpha=0.1,
                                              silent=True))],
                   final_estimator=XGBClassifier())

In [ ]:
y_pred=model2.predict(X_test)

In [ ]:
from sklearn.metrics import accuracy_score
accuracy_score(y_pred,y_test)

0.8085297418630752

In [ ]:
model2.fit(X_test,y_test)

Learning rate set to 0.091647
0:	learn: 1.0386240	total: 23.9ms	remaining: 23.9s
1:	learn: 0.9879705	total: 47.4ms	remaining: 23.6s
2:	learn: 0.9480356	total: 80.4ms	remaining: 26.7s
3:	learn: 0.9136051	total: 104ms	remaining: 25.8s
4:	learn: 0.8869357	total: 127ms	remaining: 25.2s
5:	learn: 0.8606375	total: 150ms	remaining: 24.8s
6:	learn: 0.8390780	total: 173ms	remaining: 24.5s
7:	learn: 0.8200596	total: 199ms	remaining: 24.6s
8:	learn: 0.8030695	total: 223ms	remaining: 24.6s
9:	learn: 0.7871843	total: 247ms	remaining: 24.4s
10:	learn: 0.7737284	total: 270ms	remaining: 24.2s
11:	learn: 0.7614232	total: 293ms	remaining: 24.1s
12:	learn: 0.7516837	total: 316ms	remaining: 24s
13:	learn: 0.7419573	total: 339ms	remaining: 23.9s
14:	learn: 0.7339212	total: 362ms	remaining: 23.7s
15:	learn: 0.7243619	total: 385ms	remaining: 23.7s
16:	learn: 0.7170784	total: 410ms	remaining: 23.7s
17:	learn: 0.7099760	total: 434ms	remaining: 23.7s
18:	learn: 0.7027119	total: 457ms	remaining: 23.6s
19:	learn:

StackingClassifier(cv=3,
                   estimators=[('rf',
                                RandomForestClassifier(max_depth=20,
                                                       n_estimators=500)),
                               ('cat',
                                <catboost.core.CatBoostClassifier object at 0x7fb859777dd0>),
                               ('xgb',
                                XGBClassifier(gamma=0.5, max_depth=10,
                                              min_child_weight=10,
                                              n_estimators=300, reg_alpha=0.1,
                                              silent=True))],
                   final_estimator=XGBClassifier())

In [ ]:
y_pred=model2.predict(test_X)

In [ ]:
type(y_pred)

numpy.ndarray

In [ ]:
y_pred[:10]

array([0, 0, 0, 1, 0, 0, 1, 1, 1, 0])

In [ ]:
predictions=pd.DataFrame()

In [ ]:
predictions['status_group']=y_pred.tolist()

In [ ]:
predictions.head()

status_group
0             0
1             0
2             0
3             1
4             0

In [ ]:
predictions['status_group'].value_counts()

0    9425
1    5080
2     345
Name: status_group, dtype: int64

In [ ]:
predictions=predictions.replace(to_replace=0,value='funtional')
predictions=predictions.replace(to_replace=1,value='non funtional')
predictions=predictions.replace(to_replace=2,value='funtional needs repair')

In [ ]:
predictions.head(10)

status_group
0      funtional
1      funtional
2      funtional
3  non funtional
4      funtional
5      funtional
6  non funtional
7  non funtional
8  non funtional
9      funtional

In [ ]:
left=pd.DataFrame(test['id'])
right=predictions
left['key']=range(0,len(test))
right['key']=range(0,len(test))

In [ ]:
frames=[predictions,id]
predictions=pd.merge(left,right,on='key')

In [ ]:
predictions=predictions.drop('key',axis=1)
predictions.head()

id   status_group
0  50785      funtional
1  51630      funtional
2  17168      funtional
3  45559  non funtional
4  49871      funtional

In [ ]:
predictions.to_csv(r'C:\Users\Ron\Desktop\export_dataframe.csv',index=False)